src/main/java/appeng/items/storage/ItemStorageCellBase.java

```diff
@@ -0,0 +1,271 @@
+					}
+				}
+			}
+		}
+	}
+
+	@Override
+	public int getBytes( final ItemStack cellItem )
+	{
+		return this.totalBytes;
+	}
+
+	@Override
+	public int getTotalTypes( final ItemStack cellItem )
+	{
+		return 63;
+	}
+
+	@Override
+	public boolean isBlackListed( final ItemStack cellItem, final T requestedAddition )
+	{
+		return false;
+	}
+
+	@Override
+	public boolean storableInStorageCell()
+	{
+		return false;
+	}
+
+	@Override
+	public boolean isStorageCell( final ItemStack i )
+	{
+		return true;
+	}
+
+	@Override
+	public String getUnlocalizedGroupName( final Set<ItemStack> others, final ItemStack is )
+	{
+		return GuiText.StorageCells.getUnlocalized();
+	}
+
+	@Override
+	public boolean isEditable( final ItemStack is )
+	{
+		return true;
+	}
+
+	@Override
+	public IItemHandler getUpgradesInventory( final ItemStack is )
+	{
+		return new CellUpgrades( is, 2 );
+	}
+
+	@Override
+	public IItemHandler getConfigInventory( final ItemStack is )
+	{
+		return new CellConfig( is );
+	}
+
+	@Override
+	public FuzzyMode getFuzzyMode( final ItemStack is )
+	{
+		final String fz = Platform.openNbtData( is ).getString( "FuzzyMode" );
+		try
+		{
+			return FuzzyMode.valueOf( fz );
+		}
+		catch( final Throwable t )
+		{
+			return FuzzyMode.IGNORE_ALL;
+		}
+	}
+
+	@Override
+	public void setFuzzyMode( final ItemStack is, final FuzzyMode fzMode )
+	{
+		Platform.openNbtData( is ).setString( "FuzzyMode", fzMode.name() );
+	}
+
+	@Override
+	public ActionResult<ItemStack> onItemRightClick( final World world, final EntityPlayer player, final EnumHand hand )
+	{
+		this.disassembleDrive( player.getHeldItem( hand ), world, player );
+		return new ActionResult<>( EnumActionResult.SUCCESS, player.getHeldItem( hand ) );
+	}
+
+	private boolean disassembleDrive( final ItemStack stack, final World world, final EntityPlayer player )
+	{
+		if( player.isSneaking() )
+		{
+			if( Platform.isClient() )
+			{
+				return false;
+			}
+
+			final InventoryPlayer playerInventory = player.inventory;
+			final IMEInventoryHandler inv = AEApi.instance().registries().cell().getCellInventory( stack, null, getChannel() );
+			if( inv != null && playerInventory.getCurrentItem() == stack )
+			{
+				final InventoryAdaptor ia = InventoryAdaptor.getAdaptor( player );
+				final IItemList<IAEItemStack> list = inv.getAvailableItems( getChannel().createList() );
+				if( list.isEmpty() && ia != null )
+				{
+					playerInventory.setInventorySlotContents( playerInventory.currentItem, ItemStack.EMPTY );
+
+					// drop core
+					final ItemStack extraB = ia.addItems( this.component.stack( 1 ) );
+					if( !extraB.isEmpty() )
+					{
+						player.dropItem( extraB, false );
+					}
+
+					// drop upgrades
+					final IItemHandler upgradesInventory = this.getUpgradesInventory( stack );
+					for( int upgradeIndex = 0; upgradeIndex < upgradesInventory.getSlots(); upgradeIndex++ )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)